File: flask/custom.py
# eventlet: ไลบรารีช่วยให้ Flask รองรับ WebSocket + async ได้
# docs: https://eventlet.readthedocs.io/en/latest/
import eventlet
# monkey_patch(): ดัดแปลงพฤติกรรม socket/threading ของ Python
# เพื่อให้ async ทำงานได้ โดยไม่ต้องแก้โค้ดหลัก
eventlet.monkey_patch()

from flask import Flask, render_template, request
# Flask-SocketIO: ทำให้ Flask คุยกับ client แบบ real-time ผ่าน WebSocket
# docs: https://flask-socketio.readthedocs.io/en/latest/
from flask_socketio import SocketIO

# ===== สร้าง Flask app และผูกกับ SocketIO =====
app = Flask(__name__)
socketio = SocketIO(
	app,
	async_mode='eventlet',    # ใช้ eventlet เป็น async engine
	cors_allowed_origins='*'  # เปิด CORS ให้ทุกโดเมนเข้าได้ (สะดวกเวลาทดสอบ)
) 

# ===== Built-in events =====
# connect: เรียกเมื่อ client เชื่อมต่อเข้ามา
@socketio.on('connect')
def handle_connect():
	output = {
		'message': f'ยินดีต้อนรับสู่ Flask SocketIO server ! from {request.sid} {request.remote_addr}',
		'color': '#000000'
	}
	print('Client connected', request.sid, request.remote_addr) 
	# socketio.send() → ส่งไปที่ event "message" โดยอัตโนมัติ
	socketio.send(output)

# message: เรียกเมื่อ client ใช้ socket.send()
@socketio.on('message')
def handle_message(data):
	print('Received message:', data)
	output = {
		'message': f'ข้อความที่ส่งมา: {data["message"]} from {request.sid} {request.remote_addr}',
		'color': data['color']
	}
	# ตอบกลับด้วย event "message"
	socketio.send(output)

# disconnect: เรียกเมื่อ client หลุดการเชื่อมต่อ
@socketio.on('disconnect')
def handle_disconnect():
	print('Client disconnected', request.sid, request.remote_addr)

# ===== Custom events =====
# custom_event1: client ต้องใช้ socket.emit("custom_event1", {...})
# TODO:

# custom_event2: client ต้องใช้ socket.emit("custom_event2", {...})
# TODO:

# custom_event3: client ต้องใช้ socket.emit("custom_event3", {...})
# TODO:

# ===== Route ปกติของ Flask =====
@app.route('/')
def index():
	# render หน้า custom.html (อยู่ใน templates/)
	return render_template('custom.html')

# ===== main program =====
if __name__ == '__main__':
	# ใช้ socketio.run() แทน app.run()
	socketio.run(app, host="0.0.0.0", debug=True, port=5000)

File: flask/static/custom.js
// import socket.io client แบบ ES Module
import { io } from 'https://cdn.socket.io/4.8.1/socket.io.esm.min.js';

document.addEventListener("DOMContentLoaded", function() {

	// ===== อ้างอิง element ใน HTML =====
	const connectBtn = document.getElementById("connect");
	const messageInput = document.getElementById("message");
	const sendBtn = document.getElementById("send");
	const disconnectBtn = document.getElementById("disconnect");
	const eventSelect = document.getElementById("event"); // dropdown เลือก event
	const messagesDiv = document.getElementById("messages"); // div แสดง event message
	const custom_event1Div = document.getElementById("custom_event1"); // div แสดง custom_event1
	const custom_event2Div = document.getElementById("custom_event2"); // div แสดง custom_event2
	const custom_event3Div = document.getElementById("custom_event3"); // div แสดง custom_event3
	const colorPicker = document.getElementById("color");

	// สร้าง socket object (ยังไม่เชื่อมต่อจนกว่าจะกด connect)
	const socket = io("http://localhost:5000", { autoConnect: false });

	// ===== ฟังก์ชันปรับ UI ตามสถานะการเชื่อมต่อ =====
	function UIUpdate(isConnected = false) {
		if (isConnected) {
			connectBtn.disabled = true;
			disconnectBtn.disabled = false;
			eventSelect.disabled = false;
			messageInput.disabled = false;
			sendBtn.disabled = false;
		} else {
			connectBtn.disabled = false;
			disconnectBtn.disabled = true;
			eventSelect.disabled = true;
			messageInput.disabled = true;
			sendBtn.disabled = true;
		}
	}
	UIUpdate(); // ค่าเริ่มต้น = disconnected

	// ====== Socket Events ======
	// เมื่อเชื่อมต่อสำเร็จ
	socket.on("connect", () => {
		console.log("Connected to WebSocket server");
		UIUpdate(true);
	});

	// เมื่อถูกตัดการเชื่อมต่อ
	socket.on("disconnect", () => {
		console.log("Disconnected from WebSocket server");
		UIUpdate(false);
	});

	// ====== การรับข้อความจากแต่ละ event ======
	// event message (default)
	socket.on("message", (data) => {
		console.log("Received message:", data);
		// TODO:
	});

	// event custom_event1
	// TODO:

	// event custom_event2
	// TODO:

	// event custom_event3
	// TODO:

	// ====== UI Events ======
	// ปุ่ม connect → เชื่อม socket
	connectBtn.addEventListener("click", () => {
		socket.connect();
	});

	// ปุ่ม disconnect → ตัดการเชื่อมต่อ
	disconnectBtn.addEventListener("click", () => {
		socket.disconnect();
	});

	// ปุ่ม send → ส่งข้อความไปยัง event ที่เลือกใน dropdown
	sendBtn.addEventListener("click", () => {
		const eventType = eventSelect.value;   // เลือกว่าจะส่ง event ไหน
		const message = messageInput.value;   // ข้อความที่ป้อน
		const color = colorPicker.value;      // สีที่เลือก
		if (message) {
			// ใช้ socket.emit() เพื่อส่ง event แบบกำหนดเอง
			socket.emit(eventType, { message: message, color: color });
		}
		messageInput.value = ''; // ล้าง input หลังส่ง
	});
});